### PR TITLE
Update the style guide with C++ guidelines

### PR DESCRIPTION
**Pull request by smharper**
Considering #603, we may one day start including C++ code in our source.  There are a million possible styles one can use with C++ code so I think we should try to preemptively agree on some of the big points.  This PR is essentially my proposal for a (very incomplete) C++ style.

Two points worth mentioning:
* `//` comments are preferred over `/* */`.  The justification for this (included in the docs) is that `/* */` cannot be nested so it's a real pain to comment out a bunch of code with `/* */` comments.
* There are pretty specific rules here for placing curly braces.  This is very much a matter of preference, but I think this style is preferable partly because it matches the Fortran constructs very closely (having a closing brace on its own line is very similar to an `end if` or `end subroutine`) and partly because it matches the widely-used Google C++ style guide.

This PR also makes our naming conventions more complete and specific.  I've also removed the "structures" section because it is out of date, and even if we updated it now it would probably drift out of date again.

### CODE REVIEW

**Comment by paulromano** [comment]
I'd prefer to have separate sections for Fortran, Python, and C++ rather than combining most of Fortran and C++. I also think we should adopt a modern C++ style guide (the [C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines) might fit the bill?). What do you think?

File: docs/source/devguide/styleguide.rst [hunk 1]
```diff
@@ -158,6 +163,118 @@ each side.
...
+    #define MODULE_NAME_H
+    ...
+    content
+    ...
+    #endif // MODULE_NAME_H
+
+Do not use using-directives e.g. ``using namespace foobar;``
+
+Do not use C-style casting. Always use the C++-style casts ``static_cast``,
+``const_cast``, or ``reinterpret_cast``.
```
Comment (paulromano) [inline docs/source/devguide/styleguide.rst]: This already appears in the core guidelines as ES.49

File: docs/source/devguide/styleguide.rst [hunk 2]
```diff
@@ -158,6 +163,118 @@ each side.
...
+.. code-block:: C++
+
+    #ifndef MODULE_NAME_H
+    #define MODULE_NAME_H
+    ...
+    content
+    ...
+    #endif // MODULE_NAME_H
+
+Do not use using-directives e.g. ``using namespace foobar;``
```
Comment (paulromano) [inline docs/source/devguide/styleguide.rst]: I personally think we can leave this out. The core guidelines already have a good set of practices (SF.6 and SF.7).

File: docs/source/devguide/styleguide.rst [hunk 3]
```diff
@@ -158,6 +163,118 @@ each side.
...
+lower-case with underscores (e.g. ``n_cells``) except for physics symbols that
+are written differently by convention (e.g. ``E`` for energy).
+
+Const variables should be in upper-case with underscores, e.g. ``SQRT_PI``.
+
+Curly braces
+------------
+
+For a function definition, the opening and closing braces should each be on
+their own lines.  This helps distinguish function code from the arugment list.
```
Comment (paulromano) [inline docs/source/devguide/styleguide.rst]: arugment -> argument

**Comment by smharper** [comment]
I think it's worth repeating a few of the big style points from the Core Guidelines because that entire document is quite formidable.  I want our docs to be something a new developer can read quickly and understand without too much trouble.  I know that when I first tried to read the Core Guidelines, I quickly got caught up trying to figure out what an rvalue reference and missed many of the more typical details like where whitespace goes.

Also, what do you think of the GSL referred to in the Core Guidelines?  As far as I can tell, it's intended to be all compile-time checking but it's up to the implementation if runtime checking is done instead/also.

**Comment by paulromano** [comment]
Good point -- I'm fine repeating information from the core guidelines, but if we do it would be good to put links to the relevant points so that a new developer can see the more expansive description of the rule.

On the namespace one, what we have currently means you could never do `using namespace std`; the core guidelines say you should never do this in a header (which I agree with), but I do think it can help declutter code in an implementation where there is lots of use of functionality from the standard library.

Haven't looked too much at the GSL yet so I'll make a point to check it out. What are your thoughts on it?

**Comment by smharper** [comment]
Okay, I'm convinced on the `using` directive.  I've removed that point which leaves us with the Core standard.

Regarding the GSL, I think I just need to try it out before I have a strong opinion.  However, I do think we should add some assertion macros.  MOOSE uses them, and they have helped me quickly find a bug at least once.  Plus they serve as a sort of source code level documentation.  If we decide to adopt the GSL, then our assertion macros could just wrap GSL macros.

**Comment by paulromano** [comment]
@smharper Thanks for setting the stage for C++!

**Pull request merged**